modules/can/can.c

```diff
@@ -1,56 +1,486 @@
+            systime_t ticks_to_expire = can_tx_frame_time_until_expire_X(instance->tx_mailbox[i].frame, t_now);
+            if (ticks_to_expire < min_ticks_to_expire) {
+                min_ticks_to_expire = ticks_to_expire;
+            }
+        }
+    }
+
+    struct can_tx_frame_s* frame = NULL;
+    while (can_tx_queue_iterate_I(&instance->tx_queue, &frame)) {
+        systime_t ticks_to_expire = can_tx_frame_time_until_expire_X(frame, t_now);
+        if (ticks_to_expire < min_ticks_to_expire) {
+            min_ticks_to_expire = ticks_to_expire;
+        }
+    }
+
+    worker_thread_timer_task_reschedule_I(&lpwork_thread, &instance->expire_timer_task, min_ticks_to_expire);
+}
+
+static void can_reschedule_expire_timer(struct can_instance_s* instance) {
+    chSysLock();
+    can_reschedule_expire_timer_I(instance);
+    chSysUnlock();
+}
+
+static void can_tx_frame_completed_I(struct can_instance_s* instance, struct can_tx_frame_s* frame, bool success, systime_t completion_systime) {
+    chDbgCheckClassI();
+
+    if (frame->completion_msg) {
+        struct can_transmit_completion_msg_s msg = {frame->completion_msg, success, completion_systime};
+        worker_thread_publisher_task_publish_I(&instance->tx_publisher_task, sizeof(struct can_transmit_completion_msg_s), pubsub_copy_writer_func, &msg);
+    }
+    can_tx_queue_free_I(&instance->tx_queue, frame);
+}
+
+static void can_tx_frame_completed(struct can_instance_s* instance, struct can_tx_frame_s* frame, bool success, systime_t completion_systime) {
+    if (frame->completion_msg) {
+        struct can_transmit_completion_msg_s msg = {frame->completion_msg, success, completion_systime};
+        pubsub_publish_message(&instance->tx_completion_topic, sizeof(struct can_transmit_completion_msg_s), pubsub_copy_writer_func, &msg);
+    }
+    can_tx_queue_free(&instance->tx_queue, frame);
+}
+
+static void can_expire_handler(struct worker_thread_timer_task_s* task) {
+    struct can_instance_s* instance = worker_thread_task_get_user_context(task);
+
+    // Abort expired mailboxes
+    for (uint8_t i=0; i < instance->num_tx_mailboxes; i++) {
+        chSysLock();
+        if (instance->tx_mailbox[i].state == CAN_TX_MAILBOX_PENDING && can_tx_frame_expired_X(instance->tx_mailbox[i].frame)) {
+            if (instance->driver_iface->abort_tx_mailbox_I(instance->driver_ctx, i)) {
+                instance->tx_mailbox[i].state = CAN_TX_MAILBOX_ABORTING;
+            }
+        }
+        chSysUnlock();
+    }
+
+    // Abort expired queue items
+    struct can_tx_frame_s* frame = NULL;
+    while (true) {
+        chSysLock();
+        if (!can_tx_queue_iterate_I(&instance->tx_queue, &frame) || !can_tx_frame_expired_X(frame)) {
+            chSysUnlock();
+            break;
+        }
+
+        can_tx_queue_remove_I(&instance->tx_queue, frame);
+        chSysUnlock();
+        can_tx_frame_completed(instance, frame, false, chVTGetSystemTimeX());
```

Comment:
This is a problem. can_tx_frame_completed frees frame, so it is not ok to use it in can_tx_queue_iterate_I.